utils/build_swift/presets.py

```diff
@@ -0,0 +1,247 @@
+    """
+
+    _PRESET_PREFIX = 'preset: '
+
+    def __init__(self):
+        self._parser = configparser.RawConfigParser(allow_no_value=True)
+        self._presets = {}
+
+    def _parse_raw_preset(self, section):
+        preset_name = _remove_prefix(section, PresetParser._PRESET_PREFIX)
+
+        try:
+            section_items = self._parser.items(section)
+        except configparser.InterpolationMissingOptionError as e:
+            raise MissingOptionError(e)
+
+        args, mixins = [], []
+        for (name, value) in section_items:
+            # Ignore the '--' separator, it's no longer necessary
+            if name == 'dash-dash':
+                continue
+
+            # Parse out mixin names
+            if name == 'mixin-preset':
+                mixins = list(map(lambda name: name.strip(),
+                                  value.strip().splitlines()))
+                continue
+
+            name = '--' + name  # Format as an option name
+            args.append((name, value))
+
+        return _RawPreset(preset_name, mixins, args)
+
+    def _parse_raw_presets(self):
+        for section in self._parser.sections():
+            raw_preset = self._parse_raw_preset(section)
+            self._presets[raw_preset.name] = raw_preset
+
+    def read(self, filenames):
+        """Reads and parses preset files. Throws an UnparsedFilesError if any
+        of the files couldn't be read.
+        """
+
+        parsed_files = self._parser.read(filenames)
+
+        unparsed_files = set(filenames) - set(parsed_files)
+        if len(unparsed_files) > 0:
+            raise UnparsedFilesError(list(unparsed_files))
+
+        self._parse_raw_presets()
+
+    def read_file(self, file):
+        """Reads and parses a single file.
+        """
+
+        self.read([file])
+
+    def read_string(self, string):
+        """Reads and parses a string containing preset definintions.
+        """
+
+        fp = StringIO(unicode(string))
+
+        # ConfigParser changes drastically from Python 2 to 3
+        if hasattr(self._parser, 'read_file'):
+            self._parser.read_file(fp)
+        else:
+            self._parser.readfp(fp)
+
+        self._parse_raw_presets()
+
+    def _get_preset(self, name):
+        preset = self._presets.get(name)
+        if preset is None:
+            raise PresetNotFoundError(name)
+
+        if isinstance(preset, _RawPreset):
+            preset = self._resolve_preset_mixins(preset)
+
+            # Cache resolved preset
+            self._presets[name] = preset
+
+        return preset
+
+    def _resolve_preset_mixins(self, raw_preset):
+        """Resolve all mixins in a preset, fully expanding the arguments list.
+        """
+
+        assert isinstance(raw_preset, _RawPreset)
+
+        # Expand mixin arguments
+        args = copy(raw_preset.args)
+        for mixin_name in raw_preset.mixins:
+            mixin = self._get_preset(mixin_name)
+
+            args += mixin.args
```

Comment:
```ini
[preset: base]
opt1 = 0
opt2 = 0

[preset: derived]
opt1 = 1
mixin-preset = base
opt2 = 1
```
should emit `--opt1=1 --opt1=0 --opt2=0 --opt2=1` so the derived preset can override the options.

Reply:
This was an oversight on my part. I didn't realize that it was possible to have mixins expanded in-place. I had only ever seen them declared at the beginning of a preset. I'll modify my parser to preserve this behavior and add some testing for it. Thanks!